### ISSUE TITLE
[ENH] Add fit_update() for LinearGaussianBayesianNetwork

### ISSUE THREAD

**Issue by sohamjadhav95** [pgmpy/pgmpy#2946]
## Context

This is a follow-up to #2807, which is adding Bayesian estimation
to `LinearGaussianBayesianNetwork.fit()`. Once that is implemented,
a natural next step would be supporting incremental learning with
`fit_update()`.

`DiscreteBayesianNetwork` already provides `fit_update()`, but it
works using Dirichlet pseudo-counts internally, which is specific
to the discrete case. That approach does not directly apply to
continuous networks.

For `LinearGaussianBayesianNetwork`, the conjugate prior is
Normal-Inverse-Gamma. Because of this, the update logic would need
to be implemented separately.

## What fit_update() would do

After fitting the model using `estimator="bayes"`, the posterior
parameters (μₙ, Λₙ, αₙ, bₙ) from that fit could be reused as the
prior for the next batch of data. This would allow updating the
model incrementally without refitting from scratch.

Example:

model.fit(df_batch1, estimator="bayes")
model.fit_update(df_batch2)
model.fit_update(df_batch3)

Under the Normal-Inverse-Gamma conjugacy, the result after each
update should be mathematically identical to fitting the model on
all data together.

## Implementation idea

The sufficient statistics from the previous Bayesian fit could be
stored inside each `LinearGaussianCPD`. When `fit_update()` is
called, it would read those statistics, apply the NIG update
equations with the new batch of data, and update the stored
parameters.

If `fit_update()` is called before a Bayesian fit, it should raise
a clear error.

## Relation to existing work

- Depends on #2807 being merged first
- Similar in spirit to `fit_update()` in `DiscreteBayesianNetwork`,
  but requires a separate implementation because the prior
  structure is different

I'd like to work on this if it can be assigned.

cc @ankurankan

**Comment by ankurankan** [comment]
Can you rewrite the issue in your own words and not copy-paste LLM-generated text? I need to have an understanding of how much of the issue you understand to be able to help.

**Comment by sohamjadhav95** [comment]
Ok @ankurankan,

After Looking into codebase, I have noticed that the problem is `fit()` in `LinearGaussianBayesianNetwork` only does MLE right now. Even after #2807 adds Bayesian estimation, there is still no way to update the model with new data arrives later.

`DiscreteBayesianNetwork` has `fit_update() `which uses Dirichlet pseudo-counts (fake prior observations that represent what the model already learned) to update without refitting from scratch.

It can't just be copied over because the continuous case uses Normal-Inverse-Gamma conjugate prior, not Dirichlet. Different math and different implementation needed here.

The idea is `fit_update()` would be After a Bayesian fit, store the posterior parameters on each CPD object. When new data arrives, use those stored values as the starting prior and run the update. The result is mathematically the same as refitting on all data combined.

**Comment by sohamjadhav95** [comment]
Thanks for assignment, I'll Share the implementation plan shortly before writing any code.

**Comment by sohamjadhav95** [comment]
To properly test `fit_update()`, the model first needs to be fitted with `estimator="bayes"` so the NIG parameters are populated. Since #2807 is handling that part, I see two possible ways to approach testing.
1. Add a minimal `fit(estimator="bayes")` implementation in this branch as well, so `fit_update()` can be fully tested and the PR stays self-contained. In that case I would coordinate attribute names with @Anushka-2406 to avoid conflicts when both PRs merge.
2. Manually seed the NIG parameters inside the tests and test `fit_update()` in isolation without relying on the full Bayesian fit. This would keep the scope narrow, but the complete end to end flow wouldn't be tested.

@ankurankan 
Which approach would you prefer or which seems best?

**Comment by ankurankan** [comment]
@sohamjadhav95 What does NIG mean here? Can you show me the mathematical details of how this `fit_update` method would work?

**Comment by sohamjadhav95** [comment]
@ankurankan 

NIG refers to Normal - Inverse - Gamma Distribution here, which is the conjugate prior for the parameters of a linear gaussian regression model.
For each node with parents, the local conditional model can be written as: `y = X @ β + ε, ε ~ N(0, σ²)`.  Where y is the node values, X is design matrix [1, parent_1, parent_2, ...........], β are the coefficients.

The prior over (β, σ²) is a Normal - Inverse - Gamma: `p(β, σ²) = NIG(μ₀, Λ₀, α₀, b₀)`

After observing data (X, y), the posterior parameters can be updated in closed form:
```
Λₙ = Λ₀ + XᵀX
μₙ = Λₙ⁻¹ (Λ₀ μ₀ + Xᵀ y)
αₙ = α₀ + n/2
bₙ = b₀ + ½ (yᵀy + μ₀ᵀΛ₀μ₀ − μₙᵀΛₙμₙ)
```
The CPD parameters stored in the model correspond to the MAP estimates: 
```
β_map = μₙ 
σ_map = sqrt(bₙ / (αₙ + 1))
```

The idea behind fit_update() is that after an initial Bayesian fit, the posterior parameters (μₙ, Λₙ, αₙ, bₙ) are stored with the CPD. When new data arrives, these posterior parameters are treated as the new prior (μ₀, Λ₀, α₀, b₀) and the same update equations are applied again with the new (X, y).

This gives an incremental update that is mathematically equivalent to fitting the model on the combined dataset.

**Comment by sohamjadhav95** [comment]
Hey @ankurankan,
Just checking in, happy to proceed with either approach whenever you have a moment to advise.

**Comment by ankurankan** [comment]
But if you use the Normal inverse gamma distribution, how do you do updates? If you have a given LinearGaussianBN, the distributions are conditional Gaussian. To do the fit update on this model, this conditional Gaussian would need to be the prior, which then gets updated with new data.

**Comment by sohamjadhav95** [comment]
That's a good point @ankurankan and you're right, I think I overcomplicated it.

Looking at how  `DiscreteBayesianNetwork.fit_update()` works, it doesn't require a Bayesian fit first, it just takes the current CPD values and converts them into pseudo counts using `n_prev_samples`.

For LinearGaussianBN, the equivalent idea would be to store sufficient statistics `(XᵀX, Xᵀy, yᵀy, n)` during `fit()`, and then use them in `fit_update()` by accumulating the new batch’s statistics on top and recomputing (β, σ²) from the combined totals. That's the way `fit_update()` works regardless of whether the initial fit was Bayesian or MLE and result is always equivalent to fitting on all data combined.

Is that the direction you had in mind?

**Comment by ankurankan** [comment]
What are `X` and `y` here? I think we will have to do some kind of update over the joint distribution of all variables and then recompute the LinearCPD parameters. I don't think there is a closed form solution to using conditional gaussian distributions as prior.

**Comment by sohamjadhav95** [comment]
@ankurankan,

I think I am understanding your point now, the right approach is to updating the joint Gaussian makes more sense than treating each node as a separate regression.

After `fit()`, we can store the sample mean and covariance matrix of all variables on the model. Then in `fit_update()`, instead of working with local CPDs directly, we update the joint distribution using the pooled mean and covariance formula with the new batch.

Once we have the updated joint Gaussian, we can recompute the CPD parameters from it using the standard conditional Gaussian relationships, regression coefficients from cross covariance with parents and noise variance from the conditional variance.

This way, we are effectively doing an update over the joint distribution and then re-deriving the Linear CPDs, which aligns with what you mentioned.

Also, since this only depends on the joint statistics, it should work regardless of how the initial `fit()` was done, and the result should be equivalent to fitting on all data combined.

**Comment by sohamjadhav95** [comment]
Hey @ankurankan,
I am looking for your opinion regarding the approach so we can move forward to work on the issue.

**Comment by ankurankan** [comment]
@sohamjadhav95 Yes, looks good. Would suggest writing up the mathematical equations first for the operations.

**Comment by sohamjadhav95** [comment]
Ok @ankurankan, here are the mathematical equations for `fit_update()`:

**Storing sufficient statistics after `fit()`:**
After fitting on n samples, store on the model:
- `n_samples = n`
- sample_mean μ = (1/n) Σᵢ xᵢ
- sample_cov Σ = (1/n) Σᵢ (xᵢ - μ)(xᵢ - μ)ᵀ   (MLE, ddof=0)

**Updating the joint Gaussian in fit_update():**
Given stored stats (n₁, μ₁, Σ₁) and new batch (n₂, μ₂, Σ₂):
Combined mean: μ = (n₁μ₁ + n₂μ₂) / (n₁ + n₂)
Combined covariance: Σ = [ n₁Σ₁ + n₂Σ₂ + n₁(μ₁-μ)(μ₁-μ)ᵀ + n₂(μ₂-μ)(μ₂-μ)ᵀ ] / (n₁ + n₂)

**Re-extracting CPD parameters from updated joint:**
For root node Xᵢ (no parents):
β₀ = μ[i]
σ = sqrt(Σ[i,i] · n_total / (n_total - 1))

For node Xᵢ with parents pa(Xᵢ), let k = 1 + |pa|:
β = Σ[pa,pa]⁻¹ · Σ[pa,i]
β₀ = μ[i] - β · μ[pa]
σ = sqrt( (Σ[i,i] - Σ[i,pa] · Σ[pa,pa]⁻¹ · Σ[pa,i]) · n_total / (n_total - k) )

The scaling factor n_total / (n_total - k) keeps the variance consistent with pgmpy’s default `std_estimator="unbiased"` used in `fit()`.

This way, `fit_update()` updates the joint distribution and then recomputes the CPDs, making it equivalent to fitting on all data combined.

Is this looks ideal, confirm so I can start working on code.

**Comment by ankurankan** [comment]
@sohamjadhav95 Looks good.

**Comment by sohamjadhav95** [comment]
Hey @ankurankan,
I have raised the PR, check it out and I'll be happy to receive your feedback.
